packages/puppeteer-core/src/bidi/core/Request.ts

```diff
@@ -0,0 +1,116 @@
+/**
+ * @license
+ * Copyright 2024 Google Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type * as Bidi from 'chromium-bidi/lib/cjs/protocol/protocol.js';
+
+import {EventEmitter} from '../../common/EventEmitter.js';
+
+import type BrowsingContext from './BrowsingContext.js';
+
+/**
+ * @internal
+ */
+export default class Request extends EventEmitter<{
+  // Emitted whenever a redirect is received.
+  redirect: Request;
+  // Emitted when when the request succeeds.
+  success: Bidi.Network.ResponseData;
+  // Emitted when when the request errors.
+  error: string;
+}> {
+  readonly #context: BrowsingContext;
+  readonly #event: Bidi.Network.BeforeRequestSentParameters;
+
+  #response?: Bidi.Network.ResponseData;
+  #redirect?: Request;
+  #error?: string;
+
+  constructor(
+    context: BrowsingContext,
+    event: Bidi.Network.BeforeRequestSentParameters
+  ) {
+    super();
+    this.#context = context;
+    this.#event = event;
+
+    const session = this.#session;
+    session.on('network.beforeRequestSent', event => {
+      if (event.context !== this.id) {
+        return;
+      }
+      if (event.request.request !== this.id) {
+        return;
+      }
+      if (this.#redirect) {
+        // XXX: Separate descendant redirect events?.
```

Comment:
Could you add more details what it means?

Reply:
This was used to help with calculating redirect chaining, but I think we can remove this since we can just attach listeners onto each subsequent redirect to calculate the chain in Puppeteer.